src/sentry/models/notificationaction.py

```diff
@@ -142,6 +149,39 @@ class Meta:
         db_table = "sentry_notificationactionproject"
 
 
+class ActionRegistration(metaclass=ABCMeta):
+    def __init__(self, action: NotificationAction):
+        self.action = action
+
+    @abstractmethod
+    def fire(self, data: Any) -> None:
+        """
+        Handles delivering the message via the service from the action and specified data.
+        """
+        pass
+
+    def validate_action(self, data: NotificationActionInputData) -> None:
+        """
+        Optional function to provide increased validation when saving incoming NotificationActions. See NotificationActionSerializer.
+
+        :param data: The input data sent to the API before updating/creating NotificationActions
+        :raises serializers.ValidationError: Indicates that the incoming action would apply to this registration but does is not valid.
+        """
+        pass
+
+    @classmethod
+    def serialize_available(
```

Comment:
I left it as optional because if someone implements a feature that doesn't require user input (e.g. they only have one choice for receiving notifications for some alert) it might not be necessary for it to be serialized here. Returning the empty list just omits that trigger/service/target combo from the endpoint and I think that's okay.

At bare minimum, I think actions just have to be 'fire'-able.